### PR TITLE
`exporters.onnx` output names and dynamic axes fix

### PR DESCRIPTION
**Pull request by michaelbenayoun**
# What does this PR do?

This PR adds several things:

- A check that no output other than the ones specified in the `OnnxConfig` is exported.
- A model patcher that allows to filter the model outputs during the export to only have the ones specified in the `OnnxConfig`. It also handles monkey patching of model attributes if needed.
- A check that no axis is exported as dynamic when it was specified as static in the `OnnxConfig`.
- A method, called `OnnxConfig.fix_dynamic_axis` that fixes axes that were inferred to be dynamic by ONNX during the export that are actually static, this method is called right after the export.
- `DummyInputGenerator`s now supports `numpy`.

### CODE REVIEW

File: optimum/exporters/onnx/base.py [hunk 1]
```diff
@@ -123,6 +183,7 @@ class OnnxConfig(ExportConfig, ABC):
     DEFAULT_ONNX_OPSET = 11
     ATOL_FOR_VALIDATION: Union[float, Dict[str, float]] = 1e-5
     MIN_TORCH_VERSION = GLOBAL_MIN_TORCH_VERSION
+    PATCHING_SPECS: Optional[List[PatchingSpec]] = None
```
Comment (fxmarty) [inline optimum/exporters/onnx/base.py:186]: Could you add a documentation for it just above in the docstring?

**Review by fxmarty — APPROVED** [review]
This is very nice, thanks for working on it!

File: optimum/exporters/onnx/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/exporters/onnx/base.py:186]: Done

File: optimum/exporters/onnx/base.py [hunk 2]
```diff
@@ -155,20 +216,13 @@ class OnnxConfig(ExportConfig, ABC):
             }
         ),
         "semantic-segmentation": OrderedDict({"logits": {0: "batch_size", 1: "num_labels", 2: "height", 3: "width"}}),
-        "seq2seq-lm": OrderedDict(
-            {
-                "logits": {0: "batch_size", 1: "decoder_sequence_length"},
-                "encoder_last_hidden_state": {0: "batch_size", 1: "encoder_sequence_length"},
```
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/exporters/onnx/base.py:161]: No, it is not. We do not remove anything. [We are making this the default case for every seq2seq model actually](https://github.com/huggingface/optimum/pull/731/files#diff-6fb560fd316f38fac94c55202bc2ada7c167bbe0971850d3dcbf1298694122aeR702).

File: optimum/exporters/onnx/base.py [hunk 3]
```diff
@@ -64,6 +73,57 @@ class PatchingSpec:
...
+        self.patch_ops()
+        setattr(self._model, self.orig_forward_name, self.patched_forward)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.patch_ops()
+        setattr(self._model, self.orig_forward_name, self.orig_forward)
+
+    def __call__(self, *args, **kwargs):
+        if getattr(self._model, self.orig_forward_name) is self.orig_forward:
+            logger.warning("Running the non-patched model")
```
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/exporters/onnx/base.py:123]: Yes, it is the case when someone does not use the `ModelPatcher` as a context manager. 
It is more to notify someone that would not use it as intended.

File: optimum/exporters/onnx/base.py [hunk 4]
```diff
@@ -64,6 +73,57 @@ class PatchingSpec:
...
+            orig_op = spec.orig_op if spec.op_wrapper is None else spec.op_wrapper(spec.orig_op)
+            setattr(spec.o, spec.name, orig_op)
+
+    def __enter__(self):
+        self.patch_ops()
+        setattr(self._model, self.orig_forward_name, self.patched_forward)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.patch_ops()
+        setattr(self._model, self.orig_forward_name, self.orig_forward)
```
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/exporters/onnx/base.py:119]: Which part?

- Changing the forward allows us to control which output will be present in the exported model.
- The operator patching can be (but is not anymore) used to patch operators and modules that are problematic for the export.

File: optimum/exporters/onnx/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/exporters/onnx/base.py:126]: Makes sense!

**Review by regisss — COMMENTED** [review]
Great PR!

File: optimum/exporters/onnx/base.py [hunk 5]
```diff
@@ -64,6 +73,60 @@ class PatchingSpec:
...
+        for spec in self._patching_specs:
+            orig_op = spec.orig_op if spec.op_wrapper is None else spec.op_wrapper(spec.orig_op)
+            setattr(spec.o, spec.name, orig_op)
+
+    def __enter__(self):
+        self.patch_ops()
+        setattr(self._model, self.orig_forward_name, self.patched_forward)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.patch_ops()
```
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/exporters/onnx/base.py:121]: Good catch!

**Review by regisss — APPROVED** [review]
LGTM!
